katnip/monitors/gdbserver.py

```diff
@@ -0,0 +1,128 @@
+        self.watch_signals = list(map(lambda x: x.name, signals))
+
+    def setup(self):
+        response = self.gdbmi.write(
+            'target remote %s:%d' % (self.target_host, self.target_port),
+            timeout_sec=1,
+            raise_error_on_timeout=True
+        )
+        self.logger.debug('[GDB] Command(target.remote) %s' % json.dumps(response))
+        report = GdbMiReport(response)
+        if report.get_result() == 'done':
+            response = self.gdbmi.write('-exec-continue', raise_error_on_timeout=False)
+            self.logger.debug('[GDB] Command(-exec-continue) %s' % json.dumps(response))
+            self.is_attached = True
+        else:
+            self.logger.error(report.get_error())
+        super(GdbServerMonitor, self).setup()
+
+    def teardown(self):
+        self.gdbmi.write('detach', raise_error_on_timeout=False)
+        response = self.gdbmi.write('quit', raise_error_on_timeout=False)
+        self.logger.debug('[GDB] Command(quit) %s' % json.dumps(response))
+        super(GdbServerMonitor, self).teardown()
+
+    def pre_test(self, test_number):
+        if not self.is_attached:
+            self.setup()
+        super(GdbServerMonitor, self).pre_test(test_number)
+
+    def post_test(self):
+        response = self.gdbmi.get_gdb_response(timeout_sec=self.response_timeout, raise_error_on_timeout=False)
+        self.logger.debug('[GDB] Command(alive) %s' % json.dumps(response))
+        gdb_report = GdbMiReport(response)
+        if gdb_report.is_stopped():
+            self.report.failed('The program has stopped: %s' % gdb_report.stopped_payload)
+            print(gdb_report.stopped_payload)
+            if gdb_report.stopped_payload.get('signal-name') in self.watch_signals:
+                sig_name = gdb_report.stopped_payload.get('signal-name')
+                addr = gdb_report.stopped_payload.frame.get('addr')
+                func = gdb_report.stopped_payload.frame.get('func')
+                from_target = gdb_report.stopped_payload.frame.get('from')
+                thread_id = gdb_report.stopped_payload.get('thread-id')
+                # Write report
+                self.logger.warning('Has signal: %s' % sig_name)
+                self.report.failed('Thread: %d Signal: %s Func: %s ( %d ) %s' % (
+                    thread_id, sig_name, func, addr, from_target
+                ))
+                # Write backtrace
+                response = self.gdbmi.write('bt', raise_error_on_timeout=False)
+                self.report.failed('Backtrace: %s' % GdbMiReport(response).get_console_log())
+                input('Check if need restart')
+                self.is_attached = False
+            else:
+                response = self.gdbmi.write('-exec-continue')  # continue if not interesting
+                self.logger.debug('[GDB] Command(-exec-continue) %s' % json.dumps(response))
+        super(GdbServerMonitor, self).post_test()
+
+    def _monitor_func(self):
+        sleep(0.1)
```

Comment:
add new line at the end of file

Reply:
LoL, but I have \n at the end line for each file, my Sublime make it automatically, but git was remove this line (